includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -376,185 +377,241 @@ public function payment_fields() {
-				$name                = sanitize_text_field( $order->get_billing_first_name() ) . ' ' . sanitize_text_field( $order->get_billing_last_name() );
-				$email               = sanitize_email( $order->get_billing_email() );
-				$save_payment_method = ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-new-payment-method' ] ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
-				$is_saved_method     = empty( $_POST['wcpay-payment-method'] ) && ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-payment-token' ] ); // phpcs:ignore WordPress.Security.NonceVerification.Missing
-
-				// Determine the customer making the payment, create one if we don't have one already.
-				$user        = wp_get_current_user();
-				$customer_id = $this->customer_service->get_customer_id_by_user_id( $user->ID );
-
-				if ( null === $customer_id ) {
-					// Create a new customer.
-					$customer_id = $this->customer_service->create_customer_for_user( $user, $name, $email );
-				} else {
-					// Update the existing customer with the current details. In the event the old customer can't be
-					// found a new one is created, so we update the customer ID here as well.
-					$customer_id = $this->customer_service->update_customer_for_user( $customer_id, $user, $name, $email );
-				}
+	public function process_payment_for_order( $order, $cart, $payment_information, $manual_capture, $force_save_payment_method = false ) {
+		// phpcs:ignore WordPress.Security.NonceVerification.Missing
+		$save_payment_method = ! $payment_information->is_using_saved_payment_method() && ( ! empty( $_POST[ 'wc-' . self::GATEWAY_ID . '-new-payment-method' ] ) || $force_save_payment_method );
+
+		$order_id = $order->get_id();
+		$amount   = $order->get_total();
+		$user     = $order->get_user() ?? wp_get_current_user();
+		$name     = sanitize_text_field( $order->get_billing_first_name() ) . ' ' . sanitize_text_field( $order->get_billing_last_name() );
+		$email    = sanitize_email( $order->get_billing_email() );
+		$metadata = [
+			'customer_name'  => $name,
+			'customer_email' => $email,
+			'site_url'       => esc_url( get_site_url() ),
+			'order_id'       => $order_id,
+		];
 
-				// Update saved payment method information with checkout values, as some saved methods might not have billing details.
-				if ( $is_saved_method ) {
-					try {
-						$this->customer_service->update_payment_method_with_billing_details_from_order( $payment_method, $order );
-					} catch ( Exception $e ) {
-						// If updating the payment method fails, log the error message but catch the error to avoid crashing the checkout flow.
-						Logger::log( 'Error when updating saved payment method: ' . $e->getMessage() );
-					}
-				}
+		// We only force save the card during subscriptions.
+		// TODO: This is a bit flawed; make these 2 functionalities distinct.
+		if ( $force_save_payment_method ) {
+			$metadata['payment_type'] = 'recurring';
+		}
 
```

Comment:
I need to separate the `$force_save_payment_method` flag from the metadata; those aren't really related. I'll get on that tomorrow.

Reply:
Moved to #861